src/ints/mouse_vmware.cpp

```diff
@@ -0,0 +1,247 @@
+		break;
+	}
+}
+
+static uint16_t PortReadVMware(const io_port_t, const io_width_t)
+{
+	if (reg_eax != VMWARE_MAGIC)
+		return 0;
+
+	switch (reg_cx) {
+	case VMwareCmd::GetVersion: CmdGetVersion(); break;
+	case VMwareCmd::AbsPointerData: CmdAbsPointerData(); break;
+	case VMwareCmd::AbsPointerStatus: CmdAbsPointerStatus(); break;
+	case VMwareCmd::AbsPointerCommand: CmdAbsPointerCommand(); break;
+	default:
+		LOG_WARNING("MOUSE (PS/2): unimplemented VMware command 0x%08x",
+		            reg_ecx);
+		break;
+	}
+
+	return reg_ax;
+}
+
+bool MOUSEVMWARE_NotifyMoved(const uint16_t x_abs, const uint16_t y_abs)
+{
+	auto Calculate = [](const uint16_t absolute,
+	                    int16_t &offset,
+	                    const uint16_t res,
+	                    const uint16_t clip) {
+		float unscaled; // unscaled guest mouse coordinate
+		if (mouse_video.fullscreen) {
+			// We have to maintain the diffs (offsets) between host
+			// and guest mouse positions; otherwise in case of clipped
+			// picture (like 4:3 screen displayed on 16:9 fullscreen
+			// mode) we could have an effect of 'sticky' borders if
+			// the user moves mouse outside of the guest display area
+
+			// Guest mouse position is a host mouse position +
+			// offset, which is 0 at the beginning. Once the guest
+			// mouse cursor is at the edge of the screen, and the host
+			// mouse cursor continues moving outside, the offset is
+			// increased or decreased to accomodate changes. Once
+			// the host mouse cursor starts moving back, we continue
+			// with the same offset, so that guest mouse cursor
+			// starts moving immediately.
+
+			if (absolute + offset < clip)
+				offset = static_cast<int16_t>(clip - absolute);
+			else if (absolute + offset >= res + clip)
+				offset = static_cast<int16_t>(res + clip -
+				                              absolute - 1);
+
+			unscaled = static_cast<float>(absolute + offset - clip);
+		} else {
+			// Skip the offset mechanism if not in fullscreen mode
+			unscaled = static_cast<float>(std::max(absolute - clip, 0));
+		}
+
+		return static_cast<uint16_t>(std::min(
+		        static_cast<uint32_t>(UINT16_MAX),
+		        static_cast<uint32_t>(unscaled * UINT16_MAX /
+		                                      static_cast<float>(res - 1) +
+		                              0.499)));
+	};
+
+	const auto old_x = scaled_x;
+	const auto old_y = scaled_y;
+
+	scaled_x = Calculate(x_abs, offset_x, mouse_video.res_x, mouse_video.clip_x);
+	scaled_y = Calculate(y_abs, offset_y, mouse_video.res_y, mouse_video.clip_y);
+
+	updated = true;
+
+	return mouse_vmware && (old_x != scaled_x || old_y != scaled_y);
+}
+
+void MOUSEVMWARE_NotifyPressedReleased(const uint8_t buttons_12S)
+{
+	buttons_vmware = 0;
+
+	if (bit::is(buttons_12S, b0))
+		bit::set(buttons_vmware, b5); // left button
+	if (bit::is(buttons_12S, b1))
+		bit::set(buttons_vmware, b4); // right button
+	if (bit::is(buttons_12S, b2))
+		bit::set(buttons_vmware, b3); // middle button
+
```

Comment:
How about using @kcgen's new `bit_view`? That would make this a lot less noisy.

Reply:
OK, it was possible to turn `buttons_vmware` into an `union` with a `bit_view` - as for the input parameter and `mouse.cpp`, this would require significant changes which I don't want to do before my part 2 rework gets merged (it will split `mouse.cpp` into smaller files, for easier maintenance).